### PR TITLE
Convert FxCop MessageLevel to SARIF ResultLevel equivalent, where appropriate.

### PR DESCRIPTION
**Pull request by michaelcfanning**
@lgolding @jericahuang @kschecht @Rolstenhouse

### CODE REVIEW

File: src/Sarif/Converters/FxCopConverter.cs [hunk 1]
```diff
@@ -131,13 +131,75 @@ internal Result CreateResult(FxCopLogReader.Context context)
...
+
+            result.Level = ConvertFxCopLevelToResultLevel(context.Level, out mapsDirectlyToSarifName);
+
+            if (!mapsDirectlyToSarifName)
+            {
+                // We will not recapitulate FxCop MessageLevel names (such as 
+                // "Error" and "Warning") as a property. For names that differ
+                // (such as "CriticalWarning" and "Information"), we will also 
+                // include the FxCop-specific values in the property bag.
+                TryAddProperty(result, context.Level, "Level");
```
Comment (lgolding) [inline src/Sarif/Converters/FxCopConverter.cs]: Maybe call the property "`FxCopLevel`" to avoid confusion about why there are two properties (one a first-class property, one in the property bag) both named level, and to make it clear that this is "FxCop's notion of 'level'".

**Comment by lgolding** [comment]
:+1: modulo a suggestion to rename the "Level" property in the property bag.

File: src/Sarif/Converters/FxCopConverter.cs (hunk 1 above)
Comment (michaelcfanning) [inline src/Sarif/Converters/FxCopConverter.cs]: No, sir. :) By specification, converters should not be synthesizing property names. 'Level' is the original attribute name in the FxCop project XML that contains this data and we need to preserve that.

Comment (lgolding) [inline src/Sarif/Converters/FxCopConverter.cs]: Let me think about this. The intent of the guidance for converter authors in the SARIF spec is that converters should not make guesses about the appropriate values for elements of the SARIF format that are not directly represented in the tool's native output format. Here, the converter knows exactly what the FxCop "Level" property means, and the question is how best to present it to the user in the SARIF file.

I think your concern is that if one converter decides that "FxCopLevel" is the best name for this, and another converter decides that "NativeLevel" is best, then a client could not easily consume SARIF files produced by those two different converters. That's true. On the other hand, there's nothing in the spec that requires a converter to populate a "Level" property in the property bag _at all_, so you still have the problem of converters behaving differently.

All in all, I think your argument comes down to "let's not introduce _gratuitous_ inconsistencies"; that is, "let's not do anything in our converter that another converter implementation might do differently". And I guess that preserving the tool's property name comes closest to that goal, even though (as I said) there's no requirement for a converter to emit it at all.

So bottom line, sure, let's leave it as "Level".

Comment (michaelcfanning) [inline src/Sarif/Converters/FxCopConverter.cs]: I like this simple rule, which is crystal clear and avoids lots of issues: if a converter perceives value in some tool-specific results data for which there is no equivalent SARIF slot, they can choose to populate the property bag with the information. If that data is named by the tool (via XML element or attribute, for example), the original tool name should be used, and the converter should preserve casing and values whatever is stored.

Comment (lgolding) [inline src/Sarif/Converters/FxCopConverter.cs]: Good, I'll add that guidance to the spec.

Comment (lgolding) [inline src/Sarif/Converters/FxCopConverter.cs]: Hah, actually that guidance already exists in Annex C of the spec:

> If the input data includes information for which there is no SARIF equivalent, converters may use it to populate the various property bags and tag lists defined by the SARIF format, or they may simply omit it from the output. When populating a property bag with such information, converters should use a property name that matches the name of that piece of information in the native tool format, even if that name does not conform to the camelCase convention used in the rest of this specification. This makes it easier to match these properties with the source data in the native tool format.
